libvast/src/system/importer.cpp

```diff
@@ -236,23 +238,34 @@ class driver : public importer_state::driver_base {
     CAF_IGNORE_UNUSED(path);
     // This function makes sure that we never hand out more credit than we have
     // IDs available.
-    if (desired == 0) {
+    if (desired <= 0) {
       // Easy decision if the path acquires no new credit.
       return 0;
     }
     // Calculate how much more in-flight events we can allow.
     auto& st = self_->state;
+    VAST_ASSERT(st.in_flight_slices >= 0);
     CAF_ASSERT(st.available_ids() % st.max_table_slice_size == 0);
-    auto max_credit = (st.available_ids() / st.max_table_slice_size)
-                      - st.in_flight_slices;
+    // Calculate how many table slices we can handle with available IDs.
+    auto max_available = st.available_ids() / st.max_table_slice_size;
+    // Calculate what is the numeric limit of legal credit.
+    auto max_possible = std::numeric_limits<int32_t>::max()
+                        - st.in_flight_slices;
+    // Calculate what we can hand out as new credit.
+    auto max_credit = max_available - st.in_flight_slices;
     VAST_ASSERT(max_credit >= 0);
+    // Restrict by maximum amount of credit we can reach.
+    if (max_possible < max_available)
+      return max_possible;
+    // Restruct by maximum amount of available IDs (and fetch more).
     if (max_credit <= desired) {
       // Get more IDs if we're running out.
       VAST_DEBUG(self_, "had to limit acquired credit to", max_credit);
       replenish(self_);
       st.in_flight_slices += max_credit;
```

Comment:
That is safe *iff* `replenish` doesn't mess (transitively) with `st.in_flight_slices`.